controllers/preflightvalidation_controller.go

```diff
@@ -0,0 +1,239 @@
+	log := ctrl.LoggerFrom(ctx)
+
+	modulesToCheck, err := r.getModulesToCheck(ctx, pv)
+	if err != nil {
+		return false, fmt.Errorf("failed to get list of modules to check for preflight:%w", err)
+	}
+
+	for _, module := range modulesToCheck {
+		log.Info("start module preflight validation", "name", module.Name)
+
+		verified, message := r.preflight.PreflightUpgradeCheck(ctx, &module, pv.Spec.KernelVersion)
+
+		log.Info("module preflight validation result", "name", module.Name, "verified", verified)
+
+		r.updatePreflightStatus(ctx, pv, module.Name, message, verified)
+	}
+
+	return r.checkPreflightCompletion(ctx, pv.Name, pv.Namespace)
+}
+
+func (r *PreflightValidationReconciler) getModulesToCheck(ctx context.Context, pv *kmmv1beta1.PreflightValidation) ([]kmmv1beta1.Module, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	modulesList := kmmv1beta1.ModuleList{}
+	err := r.client.List(ctx, &modulesList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get list of all Modules, %w", err)
+	}
+
+	err = r.presetNewModulesStatuses(ctx, pv, modulesList.Items)
+	if err != nil {
+		return nil, fmt.Errorf("failed to preset new modules' statuses: %w", err)
+	}
+
+	modulesStatuses := r.getModulesValidationStatusesMap(pv)
+
+	modulesToCheck := make([]kmmv1beta1.Module, 0, len(modulesList.Items))
+	for _, module := range modulesList.Items {
+		if module.GetDeletionTimestamp() != nil {
+			log.Info("Module is marked for deletion, skipping preflight validation", "name", module.Name)
+			continue
+		}
+		if status, ok := modulesStatuses[module.Name]; ok {
+			if status.VerificationStatus != kmmv1beta1.VerificationTrue {
+				modulesToCheck = append(modulesToCheck, module)
+			}
+		} else {
+			log.Info(utils.WarnString("missing status of Module CR in preflight"), "module", module.Name)
+		}
+	}
+	return modulesToCheck, nil
+}
+
+func (r *PreflightValidationReconciler) getModulesValidationStatusesMap(pv *kmmv1beta1.PreflightValidation) map[string]kmmv1beta1.CRStatus {
+	statusMap := make(map[string]kmmv1beta1.CRStatus, len(pv.Status.CRStatuses))
+	for _, status := range pv.Status.CRStatuses {
+		statusMap[status.Name] = status
+	}
+
+	return statusMap
+}
+
+func (r *PreflightValidationReconciler) updatePreflightStatus(ctx context.Context, pv *kmmv1beta1.PreflightValidation, moduleName, message string, verified bool) {
+	log := ctrl.LoggerFrom(ctx)
+	verificationStatus := kmmv1beta1.VerificationFalse
+	if verified {
+		verificationStatus = kmmv1beta1.VerificationTrue
+	}
+
+	moduleStatus := r.getModuleStatus(pv, moduleName)
+	if moduleStatus == nil {
+		log.Info(utils.WarnString("failed to find the status of Module CR in preflight"), "module", moduleName)
+		return
+	}
+	err := r.statusUpdater.PreflightSetVerificationStatus(ctx, pv, moduleStatus, verificationStatus, message)
+	if err != nil {
+		log.Info(utils.WarnString("failed to update the status of Module CR in preflight"), "module", moduleName, "error", err)
```

Comment:
Why don't we use `log.Error` here?

Reply:
Since it does not stop the flow the same way as an actual error does. In this case we continue with the loop and we continue with the processing